### DISCUSSION TITLE
Normalized Cache seems to be ignoring insert of nested query-root record

### DISCUSSION THREAD

**Discussion by EdNutting** [urql-graphql/urql#2219] [Q&A] [answered]
Hi!

I'm hoping someone can clear up my confusion. 

I have a GQL response as follows (reported by Urql dev tools) but for some reason, a record of type `content_ItemTag` does not end up in the cache's fields. 

```js
    insert_content_ItemTag: {
      returning: [
        {
          id: "23ba86ea-12f3-45f1-a5da-978efa9040e9",
          itemId: "f4422de8-00ff-4144-9a52-602215708963",
          tagId: "dc14b924-d6ae-46b3-9ab5-bff8a0e4660c",
          __typename: "content_ItemTag"
        }
      ],
      __typename: "content_ItemTag_mutation_response"
    }
```

The Urql dev tools show top-level fields for other `content_ItemTag` records (previously retrieved by queries) but after insert, and getting this response, nothing is added to the cache anywhere. I understood from the docs that with cache exchange enabled, schema awareness on, the exchange would introspect the `returning` field, identify the records of type `content_ItemTag` and process them - resulting in new ones being added  to the cache.

I am really hoping I don't have to write `updaters` just for basic cache inserts - we have 90+ tables :/ 

Really appreciate any input / thoughts / insight!

Thanks,
Ed

**Comment by EdNutting** [comment]
FWIW, I understand that even if the record were added to the cache as I expect any links to it would not be automatically inserted. I am just stuck on why no record is added at all. I have custom `resolver`s that can handle the otherwise missing links.

**Reply by kitten** [reply]
How are you checking whether a record is added currently? I'm wondering whether our GC is basically getting rid of your item since it's not referenced in queries... which, would obviously be problematic if you expected to simply query it right after 

Regarding the inserts — I have bad news, so maybe get a coffee or tea and sit down first 😅 — Graphcache indeed cannot make any assumptions about how lists change, since we cannot assume where data goes after its added. So no list will have the item added to it automatically 😅

The good news is that if you wanted to implement that, updaters are generic enough that if you're following a consistent pattern, you may be able to just write one and reuse it using a small function factory
However, it's indeed not automatic 

For inserts the code required is luckily not too involved, and usually the `info` argument, together with maybe some strategic other information you could pass, can often be enough to make an "insert updater" reusable https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#writing-links-individually

**Reply by EdNutting** [reply]
I followed up on this in the thread below.

**Comment by JoviDeCroock** [comment, marked as answer]
I mean, the root Query Field here seems highly tailored to a mutation `content_ItemTag_mutation_response` so even if it can add the following entity to the cache

```js
        {
          id: "23ba86ea-12f3-45f1-a5da-978efa9040e9",
          itemId: "f4422de8-00ff-4144-9a52-602215708963",
          tagId: "dc14b924-d6ae-46b3-9ab5-bff8a0e4660c",
          __typename: "content_ItemTag"
        }
```

It will be instantly garbage collected as it has no connection to a root-field, you would have to write a connector function for this to work, i.e. tell the cache what queryField this connects to, i.e `Query.tag(id: x) -- this entity`

**Reply by EdNutting** [reply]
Hmm, I'm not sure I understand properly. Could you explain more how the garbage collection is being applied and what you mean by a "connector function"?

**Reply by EdNutting** [reply]
(Thank you for such a rapid response!!)

**Reply by JoviDeCroock** [reply]
Every dangling entity (not connected to the root query field) is removed from the cache else you just keep adding to the memory stack, so you would have to connect it to the appropriate root-level field `Query.tag` for instance if you can query it normally with

```
query { tag }
```

So it's already removed before you can ever query it with the custom resolver 😅

**Reply by JoviDeCroock** [reply]
Generally for insertion/deletion we advise people to write an updater function, updating or derivable response bodies will be handled automatically. Linking however is something we can't assume for our users

**Reply by EdNutting** [reply]
Ok, so it determines it's not linked to any currently executing query, GCs it, and then the data is gone by the time my custom resolver goes looking for it. That's a shame!

Where should I be looking to figure out how to tell Urql "this data is connected to this other thing"? Is this a case of having to write a custom updater that searches through currently applied queries and inserts a link in the relevant places to prevent the data from being GC'd?

**Reply by JoviDeCroock** [reply]
Yes, a custom updater has you covered https://formidable.com/open-source/urql/docs/graphcache/cache-updates/#writing-links-individually

**Reply by EdNutting** [reply]
Ok..hm. Fun times. I was hoping to avoid writing custom updaters by having written a fully automatic generic resolver that can handle all the tables/fields in our schema.

My next question is going to be, is it always the case the custom resolvers don't execute during mutations or is there a more subtle condition / way to use them?

**Reply by EdNutting** [reply]
(Context: We use Hasura. I wrote a fully generic resolver that uses Proxies to intercept resolve requests and a custom auto-generated augmented schema that allows the resolver to seek across foreign keys. It handles `where`, `offset`, `limit`, etc arguments appropriately)

**Reply by JoviDeCroock** [reply]
Resolvers happen only for queries as they change the data being read, it would be a bit dangerous to apply these to subscriptions/mutations as we risk overriding data in cache that is subsequently read by queries to an unexpected format.

I mean, it's normal that we can't assume the root-connections for data, I guess the thing that mostly bit you here is the garbage collection as the heuristic is right but fights the resolver pattern I guess

**Reply by EdNutting** [reply]
The reason I ask this follow up is because I was relying on executing resolvers for computing `where` conditions. But if custom resolvers don't execute during custom updaters I can't use the same code to test if a newly inserted record needs linking into a detected active query.

**Reply by EdNutting** [reply]
Okie dokie, thank you so much for your replies. Super helpful and amazingly quick response. Really really appreciate it!

Now I know why I'm seeing what I'm seeing I think I stand a much better chance of solving this (somehow)!

**Reply by kitten** [reply]
> I wrote a fully generic resolver that uses Proxies to intercept resolve requests and a custom auto-generated augmented schema that allows the resolver to seek across foreign keys.

I personally am not a fan of using Hasura's generated and direct CRUD schemas, but I recognise its value and dev velocity, and have seen a lot of people embrace it for that. I'm not sure how many people use a Hasura + Graphcache combo (I did see a couple) but this code sounds pretty interesting to me 👀

**Reply by EdNutting** [reply]
@kitten Dev velocity and schema-driven (almost declarative) approach (particularly to permissions) were our main reasons. It's served us pretty well (if you ignore the "realtime" claims).

I'm not going to pretend this code is pretty, but here's what I've been working on:

Generic resolvers (and now I guess, updaters):
https://github.com/clowdr-app/clowdr/tree/next/subconferences/packages/shared/urql-hasura-cache-generic-resolver

Augmented schema generator (to power foreign-key lookups that are otherwise not resolvable from a standard GQL schema):
https://github.com/clowdr-app/clowdr/tree/next/subconferences/packages/node/urql-hasura-cache-schema-generator

**Reply by EdNutting** [reply]
Huh, unless I'm mistaken, there's a cheap trick to get around this. I can insert a link for the new record(s) into every query of the same field (regardless of any conditions) because the generic resolver code will filter the links anyway, thus "cleaning up" any erroneously inserted links. It should self-heal, and thus avoid needing to execute conditional evaluation (and the resolvers) during an updater.

**Reply by EdNutting** [reply]
Update: It worked 🎉

**Reply by frederikhors** [reply]
> Huh, unless I'm mistaken, there's a cheap trick to get around this. I can insert a link for the new record(s) into every query of the same field (regardless of any conditions) because the generic resolver code will filter the links anyway, thus "cleaning up" any erroneously inserted links. It should self-heal, and thus avoid needing to execute conditional evaluation (and the resolvers) during an updater.

Can you write an example, please?

**Reply by EdNutting** [reply]
@frederikhors @kitten 
This seems to do the trick: https://github.com/clowdr-app/clowdr/blob/next/subconferences/packages/shared/urql-hasura-cache-generic-resolver/src/updaters/entities.ts

But I re-emphasise for anyone else reading this thread: The _only_ reason this works is because the generic resolver I've written knows how to apply filters to data in the cache. As such it can immediately filter out any links that don't belong in the respective queries. It's also acceptable for our app because `insert`s are infrequent. If you do frequent inserts and/or have many active queries for the same data, you'll probably land in performance trouble.

**Reply by frederikhors** [reply]
Thank you.
